src/azul/queues.py

```diff
@@ -149,66 +150,104 @@ def get_queues(self, queue_names: Iterable[str]) -> Mapping[str, Queue]:
         }
 
-    def count_messages(self, queues: Mapping[str, Queue]) -> int:
+    def count_messages(self, queues: Mapping[str, Queue]) -> Mapping[str, int]:
         """
         Count the number of messages in the given queues
 
         :param queues: A collection of mapped queue_names to queues
         :return: Total count of available, in flight, and delayed messages
         """
         attribute_names = ['ApproximateNumberOfMessages' + suffix for suffix in ('', 'NotVisible', 'Delayed')]
-        total_message_count = 0
+        queue_message_count = {'total_message_count': 0}
         for queue_name, queue in queues.items():
             queue.reload()
             message_counts = [int(queue.attributes[attribute_name]) for attribute_name in attribute_names]
             queue_length = sum(message_counts)
             logger.debug('Queue %s has %i message(s) (%i available, %i in flight and %i delayed).',
                          queue_name, queue_length, *message_counts)
-            total_message_count += queue_length
-        return total_message_count
-
-    def wait_for_queue_level(self, empty: bool = True, num_bundles: int = None):
+            queue_message_count[queue_name] = queue_length
+            queue_message_count['total_message_count'] += queue_length
+        return queue_message_count
+
+    def wait_for_queue_level(self,
+                             empty: bool = True,
+                             num_bundles: int = None,
+                             min_timeout: Optional[int] = None,
+                             max_timeout: Optional[int] = None):
         """
         Wait until the total count of messages in the notify and document queues
         reaches the desired level
 
         :param empty: True to wait until the queues are empty, False to wait until not empty.
         :param num_bundles: Number of bundles being indexed (None = many bundles)
+        :param min_timeout: Specifies minimum wait time for queues to process
+                            index notifications.
+        :param max_timeout: Specifies maximum wait time for queues to process
+                            index notifications.
         """
+
+        def _compute_timeout(num_bundles):
+            # It takes approx. 6 seconds per worker to process a bundle.
+            timeout = 6 * num_bundles / config.indexer_concurrency
+            if max_timeout is not None:
+                logger.info('Maximum wait time set to %s', max_timeout)
+                timeout = min(timeout, max_timeout)
+            elif min_timeout is not None:
+                logger.info('Minimum wait time set to %s', max_timeout)
+                timeout = max(min_timeout, timeout)
+            return timeout
+
+        if max_timeout is not None and min_timeout is not None:
+            logger.error('The parameters `max_timeout` and `min_timeout` '
+                         'should not be specified simultaneously.')
+            raise Exception
```

Comment:
I don't understand why this anti-pattern keeps popping up. Unhandled exceptions are printed at the end of the process output so an unparameterized exception instance provides zero information to the user. And the log message that does may be multiple pages worth of terminal output away.

```suggestion
            raise Exception('The parameters `max_timeout` and `min_timeout` '
                         'should not be specified simultaneously.')
```